File: yaml_editor.py
# yaml_editor.py
import sys
import argparse
import yaml
import json
from jsonschema import validate, ValidationError
from pathlib import Path

# ANSI colors
COLORS = {
    'green': '\033[92m',
    'red': '\033[91m',
    'yellow': '\033[93m',
    'blue': '\033[94m',
    'reset': '\033[0m'
}

def colorize(text, color, enabled):
    return f"{COLORS[color]}{text}{COLORS['reset']}" if enabled else text

def load_yaml(content):
    try:
        return yaml.safe_load(content)
    except yaml.YAMLError as e:
        # Извлекаем позицию ошибки, если есть
        if hasattr(e, 'problem_mark'):
            mark = e.problem_mark
            line = mark.line + 1
            col = mark.column + 1
            raise ValueError(f"YAML syntax error at line {line}, column {col}: {str(e)}")
        raise ValueError(f"YAML syntax error: {str(e)}")

def load_schema(filepath):
    with open(filepath, 'r', encoding='utf-8') as f:
        return json.load(f)

def format_yaml(data, sort_keys=True):
    return yaml.dump(data, sort_keys=sort_keys, allow_unicode=True, indent=2)

def main():
    parser = argparse.ArgumentParser(description="YAML Editor & Validator")
    parser.add_argument('file', nargs='?', help='YAML file to process')
    parser.add_argument('-s', '--schema', help='JSON Schema file for validation')
    parser.add_argument('-F', '--format', action='store_true', help='Pretty-print YAML')
    parser.add_argument('-c', '--color', action='store_true', help='Force color output')
    args = parser.parse_args()

    color = args.color and sys.stdout.isatty()

    # Чтение данных
    if args.file:
        with open(args.file, 'r', encoding='utf-8') as f:
            content = f.read()
    else:
        if sys.stdin.isatty():
            print("No input provided. Pipe YAML or pass file.", file=sys.stderr)
            sys.exit(1)
        content = sys.stdin.read()

    try:
        data = load_yaml(content)
    except ValueError as e:
        print(colorize(f"❌ {e}", 'red', color), file=sys.stderr)
        sys.exit(1)

    # Валидация по схеме
    if args.schema:
        try:
            schema = load_schema(args.schema)
            validate(instance=data, schema=schema)
            print(colorize("✅ YAML is valid according to schema.", 'green', color))
        except FileNotFoundError:
            print(colorize(f"❌ Schema file not found: {args.schema}", 'red', color), file=sys.stderr)
            sys.exit(1)
        except ValidationError as e:
            # Извлекаем путь и сообщение
            path = ".".join(str(p) for p in e.path) if e.path else "root"
            print(colorize(f"❌ Validation error at {path}: {e.message}", 'red', color), file=sys.stderr)
            sys.exit(1)
    else:
        print(colorize("✅ YAML syntax is valid.", 'green', color))

    # Форматирование
    if args.format:
        print(colorize("📄 Formatted output:", 'blue', color))
        print(format_yaml(data))

if __name__ == '__main__':
    main()

File: yaml_editor.js
// yaml_editor.js
#!/usr/bin/env node
const fs = require('fs');
const path = require('path');
const yaml = require('js-yaml');
const Ajv = require('ajv');

// Colors
const colors = {
    green: '\x1b[92m',
    red: '\x1b[91m',
    yellow: '\x1b[93m',
    blue: '\x1b[94m',
    reset: '\x1b[0m'
};

function colorize(text, color, enabled) {
    return enabled ? `${colors[color]}${text}${colors.reset}` : text;
}

function loadYAML(content) {
    try {
        return yaml.load(content, { schema: yaml.JSON_SCHEMA });
    } catch (err) {
        // Извлекаем позицию, если есть
        const msg = err.message;
        const match = msg.match(/line (\d+), column (\d+)/);
        if (match) {
            throw new Error(`YAML syntax error at line ${match[1]}, column ${match[2]}: ${msg}`);
        }
        throw new Error(`YAML syntax error: ${msg}`);
    }
}

function loadSchema(file) {
    const content = fs.readFileSync(file, 'utf8');
    return JSON.parse(content);
}

function formatYAML(data) {
    return yaml.dump(data, { indent: 2, sortKeys: true });
}

function main() {
    const args = process.argv.slice(2);
    let schemaFile = null;
    let format = false;
    let color = false;
    let filePath = null;

    for (let i = 0; i < args.length; i++) {
        const arg = args[i];
        if (arg === '-s' || arg === '--schema') {
            if (i + 1 < args.length) schemaFile = args[++i];
            else { console.error('Missing schema file'); process.exit(1); }
        } else if (arg === '-F' || arg === '--format') {
            format = true;
        } else if (arg === '-c' || arg === '--color') {
            color = true;
        } else if (arg === '-h' || arg === '--help') {
            console.log(`Usage: node ${path.basename(__filename)} [options] [file]`);
            console.log('  -s, --schema <file>   JSON Schema file');
            console.log('  -F, --format          pretty-print YAML');
            console.log('  -c, --color           force color output');
            process.exit(0);
        } else {
            if (!filePath) filePath = arg;
            else { console.error(`Extra argument: ${arg}`); process.exit(1); }
        }
    }

    const useColor = color || process.stdout.isTTY;

    let content;
    if (filePath) {
        try {
            content = fs.readFileSync(filePath, 'utf8');
        } catch (err) {
            console.error(colorize(`Error reading file: ${err.message}`, 'red', useColor));
            process.exit(1);
        }
    } else {
        if (process.stdin.isTTY) {
            console.error(colorize('No input provided. Pipe YAML or pass file.', 'red', useColor));
            process.exit(1);
        }
        content = fs.readFileSync(0, 'utf8');
    }

    let data;
    try {
        data = loadYAML(content);
    } catch (err) {
        console.error(colorize(`❌ ${err.message}`, 'red', useColor));
        process.exit(1);
    }

    if (schemaFile) {
        let schema;
        try {
            schema = loadSchema(schemaFile);
        } catch (err) {
            console.error(colorize(`❌ Failed to load schema: ${err.message}`, 'red', useColor));
            process.exit(1);
        }
        const ajv = new Ajv();
        const validate = ajv.compile(schema);
        const valid = validate(data);
        if (!valid) {
            for (const err of validate.errors) {
                const path = err.instancePath || 'root';
                console.error(colorize(`❌ Validation error at ${path}: ${err.message}`, 'red', useColor));
            }
            process.exit(1);
        }
        console.log(colorize('✅ YAML is valid according to schema.', 'green', useColor));
    } else {
        console.log(colorize('✅ YAML syntax is valid.', 'green', useColor));
    }

    if (format) {
        const formatted = formatYAML(data);
        console.log(colorize('📄 Formatted output:', 'blue', useColor));
        console.log(formatted);
    }
}

if (require.main === module) {
    main();
}
